flang/include/flang/Optimizer/Dialect/Coarray/CoarrayOps.td

```diff
@@ -0,0 +1,828 @@
+//===-- CoarrayOps.td - Coarray operation definitions ------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// Definition of the Coarray dialect operations
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef FORTRAN_DIALECT_COARRAY_COARRAY_OPS
+#define FORTRAN_DIALECT_COARRAY_COARRAY_OPS
+
+include "flang/Optimizer/Dialect/Coarray/CoarrayDialect.td"
+include "flang/Optimizer/Dialect/FIRTypes.td"
+include "flang/Optimizer/Dialect/FIRAttr.td"
+include "mlir/Dialect/LLVMIR/LLVMAttrDefs.td"
+include "mlir/Dialect/LLVMIR/LLVMOpBase.td"
+include "mlir/Interfaces/LoopLikeInterface.td"
+include "mlir/IR/BuiltinAttributes.td"
+
+class coarray_Op<string mnemonic, list<Trait> traits>
+    : Op<CoarrayDialect, mnemonic, traits>;
+
+//===----------------------------------------------------------------------===//
+// Initialization and Finalization
+//===----------------------------------------------------------------------===//
+
+def coarray_InitOp : coarray_Op<"init", []> {
+  let summary = "Initialize the parallel environment";
+  let description = [{This operation will initialize the parallel environment}];
+
+  let arguments = (ins AnyIntegerType:$stat);
+}
+
+def coarray_StopOp : coarray_Op<"stop", [AttrSizedOperandSegments]> {
+  let summary = "Operation that synchronizes all execution images, clean up the"
+                "parallel runtime environment, and terminates the program.";
+  let description = [{
+    This operation synchronizes all executing images, cleans up the parallel 
+    runtime environment, and terminates the program. Calls to this operation do 
+    not return. This operation supports both normal termination at the end of a
+    program, as well as any STOP statements from the user source code.
+  }];
+
+  let arguments = (ins Optional<BoolLike>:$quiet,
+      Optional<AnyIntegerType>:$stop_code_int,
+      Arg<Optional<AnyRefOrBoxType>, "", [MemWrite]>:$stop_code_char,
+      UnitAttr:$is_error);
+
+  let hasVerifier = 1;
+}
+
+def coarray_FailImageOp : coarray_Op<"fail_image", [NoMemoryEffect]> {
+  let summary = 
+    "Causes the executing image to cease participating in program execution";
+}
+
+//===----------------------------------------------------------------------===//
+// Image Queries
```

Comment:
The STOP and FAIL IMAGE statements and intrinsics NUM_IMAGES, STOPPED_IMAGES, FAILED_IMAGES and IMAGE_STATUS also have nothing to do with coarrays. Neither does SYNC ALL, nor (arguably) all the TEAM support routines.

Perhaps this dialect should be renamed to the multi-image dialect? That name could encompass all the features related to multi-image parallel execution, of which coarrays are an important subset.